src/Kestrel.Core/Internal/Http2/HPack/HPackEncoder.cs

```diff
@@ -24,20 +24,29 @@ public bool BeginEncode(int statusCode, IEnumerable<KeyValuePair<string, string>
             _enumerator.MoveNext();
 
             var statusCodeLength = EncodeStatusCode(statusCode, buffer);
-            var done = Encode(buffer.Slice(statusCodeLength), out var headersLength);
+            var done = Encode(buffer.Slice(statusCodeLength), throwIfNoneEncoded: false, out var headersLength);
             length = statusCodeLength + headersLength;
 
             return done;
         }
 
         public bool Encode(Span<byte> buffer, out int length)
+        {
+            return Encode(buffer, throwIfNoneEncoded: true, out length);
+        }
+
+        private bool Encode(Span<byte> buffer, bool throwIfNoneEncoded, out int length)
         {
             length = 0;
 
             do
             {
                 if (!EncodeHeader(_enumerator.Current.Key, _enumerator.Current.Value, buffer.Slice(length), out var headerLength))
                 {
+                    if (length == 0 && throwIfNoneEncoded)
+                    {
+                        throw new HPackEncodingException(CoreStrings.HPackErrorNotEnoughBuffer);
```

Comment:
I know we don't plan for this to be permanent, but I would still like a test verifying this gets thrown from the first body write.

Reply:
Added